extensions/ql-vscode/src/queries-panel/query-pack-discovery.ts

```diff
@@ -0,0 +1,81 @@
+import { basename, dirname } from "path";
+import { CodeQLCliServer } from "../codeql-cli/cli";
+import { Event } from "vscode";
+import { App } from "../common/app";
+import { getOnDiskWorkspaceFolders } from "../helpers";
+import { dbSchemeToLanguage } from "../common/query-language";
+import { FALLBACK_QLPACK_FILENAME, QLPACK_FILENAMES } from "../pure/ql";
+import { FilePathDiscovery } from "./file-path-discovery";
+import { isEqualOrHigherPath } from "../pure/helpers-pure";
+
+export interface QueryPack {
+  path: string;
+  language: string | undefined;
+}
+
+/**
+ * Discovers all query packs contained in the workspace.
+ */
+export class QueryPackDiscovery extends FilePathDiscovery<QueryPack> {
+  constructor(app: App, private readonly cliServer: CodeQLCliServer) {
+    super("Query Pack Discovery", "**/{qlpack.yml,codeql-pack.yml}", app);
+  }
+
+  public get onDidChangeQueryPacks(): Event<void> {
+    return this.onDidChangePathsEmitter.event;
+  }
+
+  public findQueryPack(queryPath: string): QueryPack | undefined {
+    // Find all packs in a higher directory than the query
+    const packs = this.paths.filter((queryPack) =>
+      isEqualOrHigherPath(dirname(queryPack.path), queryPath),
+    );
+
+    // Sort by descreasing path length to find the pack nearest the query
+    packs.sort((a, b) => b.path.length - a.path.length);
+
+    // If the first two packs are from the same directory then look at the filenames
+    if (
+      packs.length >= 2 &&
+      dirname(packs[0].path) === dirname(packs[1].path)
+    ) {
+      if (basename(packs[0].path) === FALLBACK_QLPACK_FILENAME) {
+        return packs[0];
+      } else {
+        return packs[1];
+      }
+    } else {
+      return packs[0];
+    }
+  }
+
+  protected async getDataForPath(path: string): Promise<QueryPack> {
+    const language = await this.determinePackLanguage(path);
+    return { path, language };
+  }
+
+  private async determinePackLanguage(
+    path: string,
+  ): Promise<string | undefined> {
+    const packInfo = await this.cliServer.resolveLibraryPath(
```

Comment:
This works when called on query packs inside the QL repo itself, but on other repos (e.g. `github/codeql-action`) I'm getting errors like the following:
```
Resolving library paths failed with args:
    resolve library-path -v --log-to-stderr --format json --query /Users/robertbrignull/github/codeql-action/queries/qlpack.yml --additional-packs /Users/robertbrignull/github/codeql-action
[2023-06-08 16:45:56] This is codeql execute cli-server -v --log-to-stderr
[2023-06-08 16:45:57] [DETAILS] resolve library-path> Resolving query at normalized path /Users/robertbrignull/github/codeql-action/queries/qlpack.yml.
[2023-06-08 16:45:57] [DETAILS] resolve library-path> Found enclosing pack 'codeql-action-custom-queries-javascript' at /Users/robertbrignull/github/codeql-action/queries.
[2023-06-08 16:45:57] [DETAILS] resolve library-path> Adding compilation cache at /Users/robertbrignull/.codeql/compile-cache.
[2023-06-08 16:45:57] [DETAILS] resolve library-path> Resolving library dependencies from /Users/robertbrignull/github/codeql-action/queries/qlpack.yml.
[2023-06-08 16:45:57] [SPAMMY] resolve library-path> [INCOMPATIBILITY] codeql-action-custom-queries-javascript: not 0.0.0 {root: codeql-action-custom-queries-javascript@0.0.0}
[2023-06-08 16:45:57] [SPAMMY] resolve library-path> [DERIVATION] codeql-action-custom-queries-javascript: 0.0.0 {codeql-action-custom-queries-javascript: not 0.0.0 {root: codeql-action-custom-queries-javascript@0.0.0}}
[2023-06-08 16:45:57] [SPAMMY] resolve library-path> [INCOMPATIBILITY] codeql-action-custom-queries-javascript: * [*], codeql-javascript: not * {dependency: codeql-action-custom-queries-javascript@* [*] requires codeql-javascript@*}
[2023-06-08 16:45:57] [SPAMMY] resolve library-path> [DECISION 1] codeql-action-custom-queries-javascript: 0.0.0
[2023-06-08 16:45:57] [SPAMMY] resolve library-path> [DERIVATION] codeql-javascript: * {codeql-action-custom-queries-javascript: * [*], codeql-javascript: not * {dependency: codeql-action-custom-queries-javascript@* [*] requires codeql-javascript@*}}
[2023-06-08 16:45:57] [SPAMMY] resolve library-path> [INCOMPATIBILITY] codeql-javascript: * [*] {pack not found: codeql-javascript}
[2023-06-08 16:45:57] [SPAMMY] resolve library-path> [CONFLICT] codeql-javascript: * [*] {pack not found: codeql-javascript}
[2023-06-08 16:45:57] [SPAMMY] resolve library-path>   [CONSTRAINT] codeql-action-custom-queries-javascript: 0.0.0
[2023-06-08 16:45:57] [SPAMMY] resolve library-path>   [CONSTRAINT] codeql-javascript: *
[2023-06-08 16:45:57] [SPAMMY] resolve library-path>   [SATISFIER] derivation: codeql-javascript: *
[2023-06-08 16:45:57] [SPAMMY] resolve library-path>   [ALMOST_SATISFIER] <none>
[2023-06-08 16:45:57] [SPAMMY] resolve library-path> [INCOMPATIBILITY] codeql-action-custom-queries-javascript: * [*] {conflict}
[2023-06-08 16:45:57] [SPAMMY] resolve library-path> [FAILURE]
[2023-06-08 16:45:58] [SPAMMY] resolve library-path>   Because 'codeql-action-custom-queries-javascript' depends on 'codeql-javascript@*', but pack 'codeql-javascript' was not found, version solving failed.
[2023-06-08 16:45:58] [ERROR] resolve library-path> ERROR: Referenced pack 'codeql-javascript' was not found locally, and no lock file is present for 'codeql-action-custom-queries-javascript'. Run 'codeql pack install' to create the lock file and download the dependencies. (/Users/robertbrignull/github/codeql-action/queries/qlpack.yml:1,1-1)
[2023-06-08 16:45:58] [ERROR] resolve library-path> ERROR: No valid pack solution found:
                              Because 'codeql-action-custom-queries-javascript' depends on 'codeql-javascript@*', but pack 'codeql-javascript' was not found, version solving failed.
                               (/Users/robertbrignull/github/codeql-action/queries/qlpack.yml:1,1-1)
[2023-06-08 16:45:58] Exiting with code 1
```

Reply:
Hmm, maybe this is fine, or at least unavoidable. If you open these queries in the editor normally, they don't compile, and if you try to run them on the command line they fail. This is the point of the `vscode-codeql-starter and `codespaces-codeql` repos, to provide an environment where queries can be compiled. In that case it's a shame that we can't show the query language in this case, but it's not the end of the world. So long as it works in all cases where the query could be compiled and run that's ok.